test/test_h2_upgrade.py

```diff
@@ -263,3 +263,35 @@ def test_cannot_receive_data_stream_1(self, frame_factory):
             error_code=h2.errors.ErrorCodes.STREAM_CLOSED,
         )
         assert c.data_to_send() == expected_frame.serialize()
+
+    def test_client_settings_are_applied(self, frame_factory):
+        """
+        The settings provided by the client are applied and immediately
+        ACK'ed.
+        """
+        server = h2.connection.H2Connection(client_side=False)
+        client = h2.connection.H2Connection(client_side=True)
+
+        # Get the client header data and pass it to the server.
+        header_data = client.initiate_upgrade_connection()
+        server.initiate_upgrade_connection(header_data)
+
+        # This gets complex, but here we go.
+        # RFC 7540 § 3.2.1 says that "explicit acknowledgement" of the settings
+        # in the header is "not necessary". That's annoyingly vague, but we
+        # interpret that to mean "should not be sent". So to test that this
+        # worked we need to test that the server has only sent the preamble,
+        # and has not sent a SETTINGS ack, and also that the server has the
+        # correct settings.
+        expected_frame = frame_factory.build_settings_frame(
+            server.local_settings
+        )
+        assert server.data_to_send() == expected_frame.serialize()
+
+        # We violate abstraction layers here, but I don't think defining __eq__
+        # for this is worth it. In this case, both the client and server should
+        # agree that these settings have been ACK'd, so their underlying
+        # dictionaries should be identical.
+        assert (
+            client.local_settings._settings == server.remote_settings._settings
```

Comment:
I’m slightly annoyed that if the new precondition fails, we’ll get a failure (bug in product code) rather than an error (bug in test setup).

But AFAIK, the only way pytest can throw an error is inside a fixture, and it’s not worth breaking this into a separate fixture for what’s a pretty trivial distinction.